### PR TITLE
Add a mechanism for validating meson output in tests

### PR DESCRIPTION
**Pull request by jon-turney**
Add a mechanism for validating meson output in tests.

In particular, this allows checking that `faiiling-meson` and `warning-meson` tests are actually exercising the intended code-paths and failing with the expected messages (_spoiler_: some of them aren't).

There's also some unit tests which just exist to verify the expected output occurs (e.g. `test_warning_location`), which could be moved to use this mechanism.

### CODE REVIEW

**Comment by jon-turney** [comment]
- [ ] Put the expected output in test.json rather than it's own file
- [ ] Add something to Contributing.md

**Comment by jpakkane** [comment]
The error messages contain the directory name and are thus susceptible to renamings. Please update `tools/dircondenser.py` to to also update those files.

**Comment by jpakkane** [comment]
⚔️ conflicts.

**Comment by Ericson2314** [comment]
Oh this is neat!

**Comment by dcbaker** [comment]
There's a json schema for the test format in data that needs to be updated as well.

**Comment by jon-turney** [comment]
> There's a json schema for the test format in data that needs to be updated as well.

Thanks for the reminder!  I had noted that I needed to do that, but then forgot :)

**Comment by jon-turney** [comment]
pylint failure in CI seems to be due to https://github.com/PyCQA/pylint/issues/3524

File: run_project_tests.py [hunk 1]
```diff
@@ -380,6 +381,53 @@ def run_ci_commands(raw_log: str) -> T.List[str]:
...
+
+        try:
+            how, expected = next_expected(i)
+            for actual in output.splitlines():
+                if how == "re":
+                    match = bool(re.match(expected, actual))
+                else:
+                    match = (expected == actual)
+                if match:
+                    expected = next_expected(i)
```
Comment (jon-turney) [inline run_project_tests.py:419]: how, expected

File: data/test.schema.json [hunk 2]
```diff
@@ -100,6 +101,27 @@
           "prefix"
         ]
       }
+    },
+   "stdout": {
```
Comment (dcbaker) [inline data/test.schema.json:105]: it looks like there's an extra indent at this, level. Other than that the schema update looks good. Thanks!

**Comment by jon-turney** [comment]
> There's a json schema for the test format in data that needs to be updated as well.

After rebasing, InternalTests.test_validate_json fails because #6911 didn't do that, either. Sigh.

**Comment by jpakkane** [comment]
All green. Is this considered ready to merge now?

**Comment by jon-turney** [comment]
>  Is this considered ready to merge now?

It is by me :)

**Comment by dcbaker** [comment]
merged :)

**Pull request merged**